### ISSUE TITLE
 java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.Activity.runOnUiThread(java.lang.Runnable)' on a null object reference

### ISSUE THREAD

**Issue by pratikSeven** [douglasjunior/react-native-applifecycle#1]
First of all great work with the lib really appreciated as it very use full for my project.
Just wanted to bring your attention to one issue i recently identified below are the details

How did you link to the library (autolinking, manual, rn link, cocoapods)?
-> I used autolinking for the library.

What version of React Native?
-> I am using React Native version 0.64.1.

What version of the library?
-> The version of react-native-applifecycle I'm using is 0.4.0

iOS/Android version?
-> The error occurred on Android version 14.

Did the problem happen after updating React Native?
-> No, the problem occurred without any recent updates to React Native.

Are you using the library for the first time?
-> Yes, i am using this lib for the first time.

It's a bug? Provide a link to a minimal reproduction case.

-> Yes, it's a bug. and it occurred when i tried to trigger my app life cycle from app killed state(as foreground service for backup task)

Solution: 
- > I have found the solution for this as well and i have tested that and it works
Need to add null check for `initialize()` and `invalidate()`
`if(reactApplicationContext.getCurrentActivity() != null) {
 reactApplicationContext.getCurrentActivity().runOnUiThread(() ->
 ProcessLifecycleOwner
.get()
 .getLifecycle()
.addObserver(AppLifecycleModule.this)
 );
}`

**Comment by douglasjunior** [comment]
This is strange, because here we have made tests to validate this scenario and we can't reproduce it.

Maybe it's related to the deprecated React Native version? Can you reproduce the problem with React Native 0.72, 0.73 or 0.74?

**Comment by pratikSeven** [comment]
Yes i have tested this with 0.73.2
I can see the same issue

**Comment by douglasjunior** [comment]
Can you please bring more details on how to reproduce it?

**Comment by pratikSeven** [comment]
I have a native module built with setExact Alarm service to trigger a headless service from app killed state. This is where i face this issue. If app is in foreground or in background there is no issue at all lib works perfect. For me this issue only occurred when i tried to trigger foregroundService from killed state.

**Comment by douglasjunior** [comment]
Now I see, my only concern about this is if the Lifecycle listener will be registered AFTER the activity creation, because as you told the `AppLifecycleModule.java` is created when the main activity is closed.

Let's think on the following scenario:

1. Alarm service starts in background
2. React Native creates the instance of `AppLifecycleModule.java` and call the `initialize` method, but the main activity is null, so the Lifecycle listener will be not registered
3. You tap the app Icon to open the app
4. The main Activity is created
5. **The React Native will call the `initialize` method again?**

**Comment by pratikSeven** [comment]
Extremely sorry for the delay in response.
To be very honest am not really sure if initialise is getting called twice after creation of main activity as my team wanted a quick fix i just tried solving the problem and this above mentioned solution worked for me.

- I think issue is occurring because am trying background action (using alarm manager) when app is killed and in that case activity is not visible and so it is returning null
- Let me know if you some more details from me or any other help i will be happy to do so :)

**Comment by douglasjunior** [comment]
I have tried to reproduce the problem by starting a service from AlarmManager, but in none of the cases the React Native module was started before the MainActivity.

Unfortunately, to think in a final good solution, I need to reproduce the issue first.

Any help to reproduce this problem in a new project, I would greatly appreciate it.

**Comment by AlexanderWeigand** [comment]
I get the same error when starting my app. Happens about 20% of times. Help would be very appreciated! :)
Stacktrace:

```
java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.Activity.runOnUiThread(java.lang.Runnable)' on a null object reference
initialize(AppLifecycleModule.java:49)

doInitialize(ModuleHolder.java:236)

markInitializable(ModuleHolder.java:100)

notifyJSInstanceInitialized(NativeModuleRegistry.java:124)

run(CatalystInstanceImpl.java:443)

handleCallback(Handler.java:938)

dispatchMessage(Handler.java:99)

dispatchMessage(MessageQueueThreadHandler.java:27)

loopOnce(Looper.java:211)

loop(Looper.java:300)

run(MessageQueueThreadImpl.java:228)

run(Thread.java:1012)
```

More info:
```
client.android.android_version: 12
client.android.phone_model: M2010J19SY
react native version: 0.72.10
```

Edit:
Could it have something to do with using [react-native-bootsplash](https://github.com/zoontek/react-native-bootsplash)?

**Comment by pratikSeven** [comment]
Here are the steps to reproduce the issue.
1. Create native module which will trigger a alarm service with foreground service notification when app is killed state
2. Create one headless task which will trigger `BackgroundService.start(veryIntensiveTask, options)` (react-native-background-actions)
3. Now using alarm service call this headless task when app is killed
4. This should result in crash and give the error as mentioned above.

This is what i followed and got the error.

**Comment by douglasjunior** [comment]
@pratikSeven thank you for the feedback.

For the _headless task_ you are using the [Android headless JS API](https://reactnative.dev/docs/headless-js-android)?

**Comment by pratikSeven** [comment]
Yes correct

**Comment by RikSchefferAmsterdam** [comment]
We also occasionally experience these crashes, these are the logs we see:
```
07-08 08:56:20.228 30749 30806 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.Activity.runOnUiThread(java.lang.Runnable)' on a null object reference
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at com.github.douglasjunior.reactNativeAppLifecycle.modules.AppLifecycleModule.initialize(AppLifecycleModule.java:49)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at com.facebook.react.bridge.ModuleHolder.doInitialize(ModuleHolder.java:231)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at com.facebook.react.bridge.ModuleHolder.markInitializable(ModuleHolder.java:99)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at com.facebook.react.bridge.NativeModuleRegistry.notifyJSInstanceInitialized(NativeModuleRegistry.java:102)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at com.facebook.react.bridge.CatalystInstanceImpl.lambda$initialize$3(CatalystInstanceImpl.java:420)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at com.facebook.react.bridge.CatalystInstanceImpl.$r8$lambda$i8WYkDucNZjVSuyQa9C6UtXE0mA(Unknown Source:0)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at com.facebook.react.bridge.CatalystInstanceImpl$$ExternalSyntheticLambda5.run(Unknown Source:2)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:938)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:27)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:264)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:233)
07-08 08:56:20.228 30749 30806 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:923)```

**Comment by douglasjunior** [comment]
Fixed in version 1.0.0: https://github.com/douglasjunior/react-native-applifecycle/releases/tag/1.0.0

To reproduce the problem, I needed to start the application from a secondary activity, and then start a Headless Android service.

This forced the RN to create the native modules before the MainActivity exists.

And to solve the `NullPointerException` I changed the way that the lifecycle listener is created on module initialize.
